test/common/http/http1/conn_pool_test.cc

```diff
@@ -540,5 +540,38 @@ TEST_F(Http1ConnPoolImplTest, DrainCallback) {
   dispatcher_.clearDeferredDeleteList();
 }
 
+TEST_F(Http1ConnPoolImplTest, RemoteClosePendingResponses) {
+  InSequence s;
+
+  NiceMock<Http::MockStreamDecoder> outer_decoder;
+  ConnPoolCallbacks callbacks;
+  conn_pool_.expectClientCreate();
+  Http::ConnectionPool::Cancellable* handle = conn_pool_.newStream(outer_decoder, callbacks);
+  EXPECT_NE(nullptr, handle);
+
+  NiceMock<Http::MockStreamEncoder> request_encoder;
+  Http::StreamDecoder* inner_decoder;
+  EXPECT_CALL(*conn_pool_.test_clients_[0].codec_, newStream(_))
+      .WillOnce(DoAll(SaveArgAddress(&inner_decoder), ReturnRef(request_encoder)));
+  EXPECT_CALL(callbacks.pool_ready_, ready());
+
+  EXPECT_CALL(*conn_pool_.test_clients_[0].connect_timer_, disableTimer());
+  conn_pool_.test_clients_[0].connection_->raiseEvents(Network::ConnectionEvent::Connected);
+
+  callbacks.outer_encoder_->encodeHeaders(TestHeaderMapImpl{}, true);
+
+  EXPECT_CALL(*conn_pool_.test_clients_[0].codec_, dispatch(_))
```

Comment:
It would be a more accurate test if you mocked out first receiving headers without end_stream set, then receive some data without end stream set. Then, remove close should yield a call to dispatch, and you should also check that data is actually empty before finishing with decodeData(data, true). Then I think this will more accurate test what was crashing.

Also, the title of the PR is not accurate. The response is actually complete which is why we get into this state. This can happen for an HTTP/1.0 response.